### PR TITLE
Jl/caip25 permission migration/fix locked wallet behavior

### PR DESCRIPTION
**Pull request by jiexi**
## **Description**

* Fix `wallet_getPermissions` not returning `eth_accounts` when the wallet is locked
* Fix `eth_requestAccounts` not returning accounts when existing accounts are permissions after a locked wallet is unlocked
* Combine `getPermittedAccounts` and `getPermittedAccountsSorted` in `MetaMaskController`
* Cleanup specs

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/28855?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/scripts/metamask-controller.js [hunk 1]
```diff
@@ -5164,14 +5160,15 @@ export default class MetamaskController extends EventEmitter {
   }
 
   /**
-   * Gets the permitted accounts for the specified origin. Returns an empty
-   * array if no accounts are permitted.
+   * Gets the sorted permitted accounts for the specified origin. Returns an empty
+   * array if no accounts are permitted or the wallet is locked.
```
Comment (adonesky1) [inline app/scripts/metamask-controller.js:5164]: Can we add a comment explaining why there are cases when we ignore the lock and cases when we don't?

File: app/scripts/metamask-controller.test.js [hunk 2]
```diff
@@ -836,63 +836,102 @@ describe('MetaMaskController', () => {
...
-        jest
-          .spyOn(metamaskController, 'getPermittedAccounts')
-          .mockReturnValue([]);
+          expect(
+            metamaskController.getPermittedAccounts('test.com', false),
+          ).toStrictEqual([]);
+        });
 
-        await metamaskController.getPermittedAccountsSorted('test.com');
+        it('returns some accounts if there is a CAIP-25 permission for the origin and ignoreLock is true', async () => {
```
Comment (adonesky1) [inline app/scripts/metamask-controller.test.js:863]: total nit, wording here just suggests you don't get all the accounts 😓 
```suggestion
        it('returns accounts if there is a CAIP-25 permission for the origin and ignoreLock is true', async () => {
```

File: app/scripts/metamask-controller.js (hunk 1 above)
Comment (jiexi) [reply app/scripts/metamask-controller.js]: https://github.com/MetaMask/metamask-extension/pull/28855/commits/63cb7a2ae284801e30af3957fd74b87c90e0b8d3

File: app/scripts/metamask-controller.js [hunk 3]
```diff
@@ -5164,14 +5160,18 @@ export default class MetamaskController extends EventEmitter {
   }
 
   /**
-   * Gets the permitted accounts for the specified origin. Returns an empty
-   * array if no accounts are permitted.
+   * Gets the sorted permitted accounts for the specified origin. Returns an empty
+   * array if no accounts are permitted or the wallet is locked. Returns any permitted
+   * accounts if the wallet is locked and `ignoreLock` is true. This lock bypass is needed
+   * for the eth_requestAccounts and wallet_getPermission handlers which respectively either
+   * handle locked wallet behavior internally or ignore it entirely.
```
Comment (adonesky1) [inline app/scripts/metamask-controller.js:5167]: Sorry for more nitpicking just feel like this is still confusing...
```suggestion
   * Gets the sorted permitted accounts for the specified origin. Returns an empty
   * array if no accounts are permitted or the wallet is locked. Returns any permitted
   * accounts if the wallet is locked and `ignoreLock` is true. This lock bypass is needed
   * for the `eth_requestAccounts` & `wallet_getPermission` handlers both of which  
   * return permissioned accounts to the dapp when the wallet is locked.
   * TODO: we should also return permissioned accounts on `eth_accounts` calls 
   * when the wallet is locked
```

File: app/scripts/metamask-controller.test.js [hunk 4]
```diff
@@ -836,63 +836,102 @@ describe('MetaMaskController', () => {
...
+            .spyOn(metamaskController, 'sortAccountsByLastSelected')
+            .mockResolvedValue(['not_empty']);
 
-        expect(metamaskController.getPermittedAccounts).toHaveBeenCalledWith(
-          'test.com',
-        );
+          expect(
+            metamaskController.getPermittedAccounts('test.com', true),
+          ).not.toStrictEqual([]);
+        });
```
Comment (adonesky1) [inline app/scripts/metamask-controller.test.js:883]: 🤔 are we able to make a positive assertion about what this returns?

File: app/scripts/metamask-controller.js [hunk 5]
```diff
@@ -6531,7 +6529,7 @@ export default class MetamaskController extends EventEmitter {
         method: NOTIFICATION_NAMES.unlockStateChanged,
         params: {
           isUnlocked: true,
-          accounts: await this.getPermittedAccountsSorted(origin),
+          accounts: await this.getPermittedAccounts(origin),
```
Comment (adonesky1) [inline app/scripts/metamask-controller.js:6532]:
```suggestion
          accounts: this.getPermittedAccounts(origin),
```

File: app/scripts/metamask-controller.js [hunk 6]
```diff
@@ -7117,7 +7115,7 @@ export default class MetamaskController extends EventEmitter {
               newAccounts
             : // If the length is 2 or greater, we have to execute
               // `eth_accounts` vi this method.
-              await this.getPermittedAccountsSorted(origin),
+              await this.getPermittedAccounts(origin),
```
Comment (adonesky1) [inline app/scripts/metamask-controller.js:7118]:
```suggestion
              this.getPermittedAccounts(origin),
```

File: app/scripts/metamask-controller.test.js (hunk 4 above)
Comment (jiexi) [reply app/scripts/metamask-controller.test.js:883]: overlaps with the happy case below, but yeah i suppose that would be better

Comment (jiexi) [reply app/scripts/metamask-controller.test.js]: https://github.com/MetaMask/metamask-extension/pull/28855/commits/3c8b24b34b3728b654f5d53809da768a723317ce#diff-0aea0fd14a27750ddf078dd4ef363f4d10b91285954ad5e1488edcad1d5bd306R882

**Review by adonesky1 — APPROVED** [review]
LGTM!

**Pull request merged**